fix(redshift_mcp_server): Fix missing region parameter while creation of boto3 client

HarshGrandeur opened the pull request:
…for issues - https://github.com/awslabs/mcp/issues/928

<!-- markdownlint-disable MD041 MD043 -->
Fixes

## Summary

Fix missing region parameter in boto redshift client creation. This resulted in the error `Error creating Redshift client: You must specify a region.`

### Changes

Added the missing region inside the boto redshift client.

### User experience

> Before

The user was not able to connect to Redshift mcp server due to incorrect configuration of boto redshift client.

> After

The user is able to get the results for the available redshift clusters and related queries which are present inside redshift mcp server.

#### Reference
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.client:~:text=config%20(botocore.client.Config)%20%E2%80%93%20Advanced,botocore%20config%20documentation%20for%20more%20details.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x ] I have reviewed the [contributing guidelines](https://github.com/awslabs/mcp/blob/main/CONTRIBUTING.md)
* [ x] I have performed a self-review of this change
* [ x] Changes have been tested
* [ x] Changes are documented

Is this a breaking change? N

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the [project license](https://github.com/awslabs/mcp/blob/main/LICENSE).

HarshGrandeur commented:
Fixed failing tests

grayhemp reviewed (changes requested):
@HarshGrandeur could you please provide a proper repro in the issue #928 description? What is currently in there is very vague.

Please include details about:

- Steps leading to the error
- AWS client configuration
- Environment variables.

Thank you.

HarshGrandeur commented:
> @HarshGrandeur could you please provide a proper repro in the issue #928 description? What is currently in there is very vague.
> 
> Please include details about:
> 
> * Steps leading to the error
> * AWS client configuration
> * Environment variables.
> 
> Thank you.

I have updated the issue. Please let me know if this looks better. I have added more details in the reproduction steps.

grayhemp commented:
@HarshGrandeur may I ask why you closed it?

grayhemp commented:
@HarshGrandeur, BTW, your repro in #928 doesn't seem to work. See https://github.com/awslabs/mcp/issues/928#issuecomment-3134843151.

HarshGrandeur commented:
reopening this based on : https://github.com/awslabs/mcp/issues/928#issuecomment-3176047417

HarshGrandeur commented:
@grayhemp Do you need any changes ?

grayhemp commented:
# RCA

Based on my RCA of the boto3 behavior with region specified, both the current implementation and the solution provided in this PR lead to unexpected behavior.

## Current implementation

When `AWS_PROFILE` is set, creates a session/client with the profile but no region argument.

- With no `AWS_REGION` set
  - [X] Region is in `~/.aws/config` - works as expected picking the region from the config
  - [ ] No region in `~/.aws/config` - errors with `Error creating Redshift client: You must specify a region`, the default `us-east-1` region is expected to be used
- With `AWS_REGION` set
  - [ ] Region is in `~/.aws/config` - ignores `AWS_REGION` using the region specified in the config, though `AWS_REGION` is expected to be used
  - [ ] No region in `~/.aws/config` - ignores `AWS_REGION` and errors with `Error creating Redshift client: You must specify a region`, though `AWS_REGION` is expected to be used

## @HarshGrandeur's fix

When `AWS_PROFILE` is set, creates a session/client with the profile and the region argument either as in `AWS_REGION` if the env variable is set or default `us-east-1`:

- With no `AWS_REGION` set
  - [ ] Region is in `~/.aws/config` - uses the default `us-east-1` instead of the one in `~/.aws/config`, expected to use the region set in `~/.aws/config`
  - [X] No region in `~/.aws/config` - uses the default `us-east-1` region
- With `AWS_REGION` set
  - [X] Region is in `~/.aws/config` - uses `AWS_REGION` instead of the one in `~/.aws/config`
  - [X] No region in `~/.aws/config` - uses the one set in `AWS_REGION`

Better, but still one miss.

## The problem

The problem appeared to be in boto3 not supprting `AWS_REGION` as an env variable (🤦 RTFM), see <https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html>. What it does support is `AWS_DEFAULT_REGION`.

If we use `AWS_DEFAULT_REGION` instead of `AWS_REGION`, then without modifying the current implementation (main branch) we will have the following.

When `AWS_PROFILE` is set, creates a session/client with the profile but no region argument.

- With no `AWS_DEFAULT_REGION` set
  - [X] Region is in `~/.aws/config` - works as expected picking the region from the config
  - [ ] (unclear if it's a miss, see below) No region in `~/.aws/config` - errors with `Error creating Redshift client: You must specify a region`
    - Shall it use the default `us-east-1` region?
    - Or we must require either `AWS_DEFAULT_REGION` or `AWS_REGION` or the region in `~/.aws/config` to be specified?
- With `AWS_DEFAULT_REGION` set
  - [X] Region is in `~/.aws/config` - uses `AWS_DEFAULT_REGION` instead of the one in `~/.aws/config`
  - [X] No region in `~/.aws/config` - uses the one in `AWS_DEFAULT_REGION`

In this case we have the following to address.

The one unclear behavior item (With no `AWS_DEFAULT_REGION` set, No region in `~/.aws/config`). Similarly to the note under the `AWS_DEFAULT_REGION` description in the AWS CLI doc <https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html>, it would probably be the best to expect either `AWS_DEFAULT_REGION` or `AWS_REGION` or the region in `~/.aws/config`, and let it error "officially" otherwise. We need to update the prompts with instructions to print out errors and provide suggestions on how to deal with them. That way the behavior will be compatible with AWS CLI which is the defacto standard setting expectations.

What to do with `AWS_REGION`? It can be treated similarly to the `AWS_REGION` in the AWS CLI tool (see the same doc page), which overrides the `AWS_DEFAULT_REGION` and the rest of the configuration if specified. We will need to be explicitly setting it as the session/client parameter if it is specified. There should be no default value for it.

Just a reminder, AWS CLI follows this order or precedence:

1. AWS_REGION (highest priority)
2. AWS_DEFAULT_REGION
3. Profile config region
4. Default region (if any).

We also need to carefully document the behavior including the precedence in both README.md and the prompts, as well as cover it all with tests.

What do you think? @HarshGrandeur @harishcodes

grayhemp commented:
So, the action items, if you folks aligned with the RCA/suggestion:

Code changes:

- Modify clients' creation to handle AWS_REGION explicitly as top precedence override
  - In the client creation methods
    - If `self.aws_region` is set then pass it to the boto3's `client()`
    - If not, let boto3 handle AWS_DEFAULT_REGION + profile config automatically, no code needed
- Remove hardcoded us-east-1 default - let it error if no region is configured
- Update instructions/prompts with region configuration guidance and error handling notes
- Add test cases covering all region configuration scenarios including the error case

Doc changes:

- Document region precedence order under the "AWS Client Requirements" section (as specified in my previous message)
- Add `AWS_DEFAULT_REGION` and remove the default value from `AWS_REGION` in the "Environment Variables" section
